### ISSUE TITLE
[call_tf] Error when applying to a TF function with unknown output shape restored from a SavedModel

### ISSUE THREAD

**Issue by gnecula** [google/jax#7048]
It seems that when we load a function from a SavedModel and the function has outputs of unknown shape, then `call_tf` fails with the following error:
```
  File "/Users/necula/Source/jax/jax/experimental/jax2tf/call_tf.py", line 112, in res_sig_to_aval
    return core.ShapedArray(res_sig.shape, jax2tf_internal._to_jax_dtype(res_sig.dtype))
  File "/Users/necula/Source/jax/jax/core.py", line 1043, in __init__
    self.shape = canonicalize_shape(shape)
  File "/Users/necula/Source/jax/jax/core.py", line 1400, in canonicalize_shape
    raise _invalid_shape_error(shape)
TypeError: Shapes must be 1D sequences of concrete values of integer type, got (None, 2).
```

This is the repro:
```
    import haiku as hk
    import tree
    import numpy as np
    import tensorflow as tf

    import os
    os.environ['HAIKU_FLATMAPPING'] = '0'

    def resnet(x, is_training):
      net = hk.nets.ResNet18(2)
      return net(x, is_training)

    resnet = hk.transform_with_state(resnet)
    inputs = np.random.random((5, 64, 64, 3)).astype(np.float32)
    rng = jax.random.PRNGKey(42)
    params, states = resnet.init(rng, inputs, is_training=True)
    logits, states = resnet.apply(params, states, rng, inputs, is_training=True)
    # We need to save the model with is_training=False because of batch_norm
    resnet_apply = functools.partial(resnet.apply, is_training=False)
    # Save the Haiku.apply callable as a tf.SavedModel (and the params and states)
    save_dir = '/tmp/resnet'
    # Jax -> TF
    tf_fn = jax2tf.convert(
        resnet_apply,
        polymorphic_shapes=[None, None, None, '(b, ...)'],
        with_gradient=True)
    # TF -> SavedModel
    model = tf.Module()
    model.fn = tf.function(tf_fn, autograph=False)
    model.fn.get_concrete_function(
        tree.map_structure(tf.zeros_like, params),
        tree.map_structure(tf.zeros_like, states),
        tree.map_structure(tf.zeros_like, rng),
        tf.TensorSpec((None,) + inputs.shape[1:], inputs.dtype))
    # tf.TensorSpec(inputs.shape, inputs.dtype))
    options = tf.saved_model.SaveOptions(experimental_custom_gradients=True)
    _ = tf.saved_model.save(model, save_dir, options=options)
    restored_module = tf.saved_model.load(save_dir)
    # Run inference for the restored module
    # This fails.
    jax2tf.call_tf(restored_module.fn)(params, states, rng, inputs)
```

**Comment by gnecula** [comment]
It seems that the problem is the loaded function graph contains some TF ops, e.g., `PartitionedCall` which during shape inference return unknown shape even if all the input shapes are known. For a general TF graph, this can happen, because there are TF ops whose output shape is data dependent. Is this the case with the TF ops that jax2tf generates? Perhaps.

**Comment by gnecula** [comment]
Fixed by #7119
